common/src/main/java/io/netty/util/ThreadDeathWatcher.java

```diff
@@ -121,10 +125,10 @@ public static boolean awaitInactivity(long timeout, TimeUnit unit) throws Interr
             throw new NullPointerException("unit");
         }
 
-        Thread watcherThread = ThreadDeathWatcher.watcherThread;
-        if (watcherThread != null) {
-            watcherThread.join(unit.toMillis(timeout));
-            return !watcherThread.isAlive();
+        Thread t = watcherThread.get();
+        if (t != null) {
```

Comment:
The synchronization approach used by this class is racy:

1. Watcher#run detects there are no more pending entries, so sets watcher thread to `null`.
2. schedule puts something on the `pendingEntries` queue.
3. `awaitInactivity` is called.
4. schedule or Watcher#run sets the watcher thread.

`awaitInactivity` may exit prematurely.

The comments provide a loop hole that this method should only be called after all calls to `watch`. However it makes me a bit uneasy that the code responsible for watching all threads is itself racy. I wonder if we can have a `private static void Thread TERMINATED = ...;` or some token object that is respected to avoid scheduling new work after the thread death watcher is "terminated" (e.g. `awaitInactivity` is called).